vllm/worker/cache_engine_vmm.py

```diff
@@ -0,0 +1,264 @@
+                         ParallelConfig, SchedulerConfig)
+from vllm.logger import init_logger
+from vllm.utils import (STR_DTYPE_TO_TORCH_DTYPE, TORCH_DTYPE_TO_STR_DTYPE,
+                        get_dtype_size, is_pin_memory_available)
+
+logger = init_logger(__name__)
+
+
+# TODO: maybe we need a base class for CacheEngineVMM and CacheEngine
+class CacheEngineVMM:
+    """Manages the KV cache.
+
+    This class is responsible for initializing and managing the GPU and CPU KV
+    caches. It also provides methods for performing KV cache operations, such
+    as swapping and copying.
+    """
+
+    def __init__(
+        self,
+        cache_config: CacheConfig,
+        model_config: ModelConfig,
+        parallel_config: ParallelConfig,
+        scheduler_config: SchedulerConfig,
+        device_config: DeviceConfig,
+    ) -> None:
+        self.cache_config = cache_config
+        self.model_config = model_config
+        self.parallel_config = parallel_config
+        self.scheduler_config = scheduler_config
+        self.device_config = device_config
+
+        if self.device_config.device_type != "cuda":
+            raise RuntimeError("VMM only support cuda device.")
+
+        self.head_size = model_config.get_head_size()
+        self.num_layers = model_config.get_num_layers(parallel_config)
+        self.num_kv_heads = model_config.get_num_kv_heads(parallel_config)
+
+        self.block_size = cache_config.block_size
+        self.block_bytes_size = self.cache_config.block_bytes_size
+        self.num_gpu_blocks = cache_config.num_gpu_blocks
+        self.num_cpu_blocks = cache_config.num_cpu_blocks
+
+        if cache_config.cache_dtype == "auto":
+            self.dtype = model_config.dtype
+        else:
+            self.dtype = STR_DTYPE_TO_TORCH_DTYPE[cache_config.cache_dtype]
+
+        self.dtype_size = get_dtype_size(self.dtype)
+        self.max_batch_size = self.scheduler_config.max_num_seqs
+        self.max_seq_len = self.scheduler_config.max_model_len
+        # If max_seq_len is not divisible by block_size,
+        # round up to the nearest value that is.
+        if self.max_seq_len % self.block_size != 0:
+            self.max_seq_len = ((self.max_seq_len // self.block_size + 1) *
+                                self.block_size)
+            logger.warning(
+                "self.max_seq_len mod self.block_size != 0, "
+                "round up max_seq_len to %d", self.max_seq_len)
+
+        self.token_size = self.num_kv_heads * self.head_size
+        self.sequence_buffer_size = self.max_seq_len * self.token_size
+        self.sequence_buffer_bytes_size = (self.sequence_buffer_size *
+                                           self.dtype_size)
+        self.cache_space_size = self.max_batch_size * self.sequence_buffer_size
+        self.cache_sapce_bytes_size = self.cache_space_size * self.dtype_size
+
+        assert self.cache_sapce_bytes_size % self.block_bytes_size == 0, \
```

Comment:
Thank you for your feedback.
In [vllm/engine/arg_utils.py](https://github.com/izhuhaoran/vllm/blob/suri_vllm/vllm/engine/arg_utils.py#L688), I checked 'if self.block_bytes_size % single_token_bytes_size ! = 0:', which in this model you mentioned does cause assert to be triggered here. 

Here it can be modified so that when % ! = 0, the least common multiple of single_token_bytes_size and block_bytes_size is computed as the new block_bytes_size.